### PR TITLE
Fixes updating selection and selection priorty of settlers vs buildings.

### LINKED ISSUE #430: Selections aren't updated
When you have some soldiers selected and some of them die, the selection stays the same and is not updated.

### PR DESCRIPTION
**Pull request by andreas-eberle**
Fixes #430 and #431.

### CODE REVIEW

**Comment by andreas-eberle** [comment]
@michaelzangl: Can you have a look at the changes?

File: jsettlers.logic/src/main/java/jsettlers/input/GuiInterface.java [hunk 1]
```diff
@@ -107,6 +110,14 @@ public GuiInterface(IMapInterfaceConnector connector, IGameClock clock, ITaskSch
 		this.multiplayer = multiplayer;
 		this.constructionMarksCalculator = new ConstructionMarksThread(grid.getConstructionMarksGrid(), clock, player);
 
+		this.refreshSelectionTimer = new Timer("refreshSelectionTimer");
```
Comment (michaelzangl) [inline jsettlers.logic/src/main/java/jsettlers/input/GuiInterface.java]: Please do not use a plain timer for everyting that needs updating.

Use `UIUpdater` instead.

File: jsettlers.logic/src/main/java/jsettlers/input/GuiInterface.java [hunk 2]
```diff
@@ -621,8 +632,17 @@ private void setSelection(SelectionSet selection) {
...
+		if (!currentSelection.isEmpty()) {
+			SelectionSet newSelection = new SelectionSet();
+
+			for (ISelectable selected : currentSelection) {
+				if (selected.isSelected()) {
+					newSelection.add(selected);
+				}
+			}
+
+			setSelection(newSelection);
```
Comment (michaelzangl) [inline jsettlers.logic/src/main/java/jsettlers/input/GuiInterface.java]: This is a task for the game logic - not the UI.

You have race conditions in this code.

**Comment by michaelzangl** [comment]
The best thing to do is to notify the UI that the selection changed.

Alternatively, you can let the selection panel or it's components (`SelectionRow`) implement `IUpdateReceiver<SelectionSet>` and use an `UIUpdater` to handle the changes. See the `ToolsPanel` on how to use it. Call a `forceUpdate()` to load the values - that way you can simply move all panel initialisation to the update method.

Mind that buildings (and people) can also change players - settlers need to be deselected then.

You do not need a special case handling for empty sets, simply call `removeAll()` on the root panel of your selection layout.

File: jsettlers.logic/src/main/java/jsettlers/input/SelectionSet.java [hunk 3]
```diff
@@ -61,7 +62,8 @@ public synchronized void add(ISelectable selectable) {
 		if (selectionType.priority < this.selectionType.priority) {
 			return; // selectable is of lower priority
 		} else if (selectionType.priority > this.selectionType.priority) {
-			clear();
+			setSelected(false);
+			this.set = new ArrayList<>();
```
Comment (michaelzangl) [inline jsettlers.logic/src/main/java/jsettlers/input/SelectionSet.java]: set.clear() ?

Comment (andreas-eberle) [inline jsettlers.logic/src/main/java/jsettlers/input/SelectionSet.java]: what do you mean?

File: jsettlers.logic/src/main/java/jsettlers/input/GuiInterface.java (hunk 1 above)
Comment (andreas-eberle) [inline jsettlers.logic/src/main/java/jsettlers/input/GuiInterface.java]: As this is in the jsettlers.logic project, we shoulnd't use UI stuff.

File: jsettlers.logic/src/main/java/jsettlers/input/GuiInterface.java (hunk 2 above)
Comment (andreas-eberle) [inline jsettlers.logic/src/main/java/jsettlers/input/GuiInterface.java]: It actually is in the game logic.

File: jsettlers.logic/src/main/java/jsettlers/input/SelectionSet.java (hunk 3 above)
Comment (michaelzangl) [inline jsettlers.logic/src/main/java/jsettlers/input/SelectionSet.java]: this.set.clear() does not allocate a new list.

**Comment by andreas-eberle** [comment]
@michaelzangl: Is it ok for you now?

**Comment by andreas-eberle** [comment]
As discussed with @michaelzangl in chat, PR is accepted.

**Pull request merged**